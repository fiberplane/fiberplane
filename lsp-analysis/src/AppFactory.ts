import { Hono } from "hono";
import type { RouteElement, RouteTree } from "./types";

export class AppFactory {
  private trees: Array<RouteTree>;
  private apps: Map<string, Hono> = new Map();
  private history: Map<string, RouteElement>;

  constructor(trees: Array<RouteTree>) {
    this.trees = trees;
  }

  public resetHistory() {
    this.history = new Map();
  }

  public hasVisited(id: string): boolean {
    return this.history.has(id);
  }

  public getHistoryLength(): number {
    return this.history.size;
  }

  public getVisitedNode(id: string) {
    return this.trees.find((tree) => tree.id === id);
  }

  public getApp(entryId: string): Hono {
    const root = this.trees.find((tree) => tree.id === entryId);
    this.resetHistory();

    if (!root) {
      throw new Error("Could not find root tree");
    }

    return this.createApp(root);
  }

  private createApp(tree: RouteTree) {
    const app = new Hono();

    this.apps.set(tree.id, app);

    app.use((c, next) => {
      this.history.set(tree.id, tree);
      return next();
    });

    // for (const entry of tree.entries) {
    //   switch (entry.type) {
    //     case "ROUTE_ENTRY": {
    //       app.get(entry.path, (c) => {
    //         this.history.set(entry.id, entry);
    //         return c.text("Ok");
    //       });
    //       continue;
    //     }

    //     case "ROUTE_TREE_REFERENCE": {
    //       if (this.apps.has(entry.targetId)) {
    //         const targetApp = this.apps.get(entry.targetId);
    //         app.route(entry.path, targetApp);
    //       } else {
    //         const tree = this.trees.find((tree) => tree.id === entry.targetId);
    //         if (tree) {
    //           const targetApp = this.createApp(tree);
    //           app.route(entry.path, targetApp);
    //         }
    //       }
    //       continue;
    //     }

    //     case "MIDDLEWARE_ENTRY": {
    //       const middlewareFunc = async (c, next) => {
    //         this.history.set(entry.id, entry);
    //         await next();
    //       };
    //       if (entry.path) {
    //         app.use(entry.path, middlewareFunc);
    //       } else {
    //         app.use(middlewareFunc);
    //       }
    //       continue;
    //     }

    //     default: {
    //       const exhaustiveCheck: never = entry;
    //       // Typescript should never happen
    //       throw new Error(`Unsupported entry type: ${exhaustiveCheck}`);
    //     }
    //   }
    // }

    return app;
  }
}
