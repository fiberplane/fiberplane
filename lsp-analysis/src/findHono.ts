import type {
  ModuleReference,
  RouteEntry,
  RouteTree,
  SourceReference,
  TsArrowFunction,
  TsLanguageService,
  TsModuleResolutionHost,
  TsNode,
  TsProgram,
  TsReferenceEntry,
  TsSourceFile,
  TsType,
} from "./types";

export function findHonoRoutes(
  server: TsLanguageService,
  ts: TsType,
): {
  errorCount?: number;
  results: Array<RouteTree>;
} {
  const program = server.getProgram();
  if (!program) {
    throw new Error("Program not found");
  }

  let errorCount = 0;
  const checker = program.getTypeChecker();

  const apps: Array<RouteTree> = [];

  const files = program.getSourceFiles();
  const fileMap: Record<string, TsSourceFile> = {};
  for (const file of files) {
    fileMap[file.fileName] = file;
  }

  for (const sourceFile of files) {
    if (!sourceFile.isDeclarationFile) {
      ts.forEachChild(sourceFile, (file: TsSourceFile) =>
        visit(file, file.fileName),
      );
    }
  }

  function visit(node: TsNode, fileName: string) {
    if (ts.isVariableDeclaration(node) && node.initializer) {
      const type = checker.getTypeAtLocation(node.initializer);
      const typeName = checker.typeToString(type);
      if ("intrinsicName" in type && type.intrinsicName === "error") {
        errorCount++;
        console.error("Error in type check");
        console.error("In: ", node.getSourceFile().fileName, node.kind);
        console.error("Ndoe text:", node.getFullText());
        console.error("type information", type.getSymbol());
      }

      if (typeName.startsWith("Hono<")) {
        const honoInstanceName = node.name.getText();

        const current: RouteTree = {
          name: honoInstanceName,
          fileName: node.getSourceFile().fileName,
          entries: [],
        };

        findRoutes(current, node.parent.parent.parent);
        apps.push(current);
      }
    }

    ts.forEachChild(node, (child) => visit(child, fileName));
  }

  function findRoutes(routeTree: RouteTree, scope: TsNode) {
    const honoInstanceName = routeTree.name;
    function visit(node: TsNode) {
      if (
        ts.isCallExpression(node) &&
        ts.isPropertyAccessExpression(node.expression)
      ) {
        const methodName = node.expression.name.getText();
        const objectName = node.expression.expression.getText();

        if (
          objectName === honoInstanceName &&
          ["put", "get", "post", "all"].includes(methodName)
        ) {
          const [firstArgument, ...args] = node.arguments;
          const entry: RouteEntry = {
            method: methodName,
            path: JSON.parse(firstArgument.getText()),
            sources: [],
          };

          // Add the tree node to the list of entries
          // Later the entry will be filled with source references
          routeTree.entries.push(entry);

          for (const arg of args) {
            if (ts.isArrowFunction(arg)) {
              // const source = extractArrowFunctionContext(arg, ts);
              const source = extractReferencesFromArrowFunction(arg, program, ts);
              // console.log('arg.content', arg.getText())
              if (source) {
                entry.sources.push(...source);
              }
              // const sourceFile = arg.getSourceFile();
              // const position = sourceFile.getLineAndCharacterOfPosition(
              //   arg.getStart(),
              // );
              // entry.sources.push({
              //   character: position.character,
              //   line: position.line,
              //   fileName: sourceFile.fileName,
              //   content: arg.getText(),
              //   references: [],
              //   modules: {},
              // });
              // console.log('entry.sources', entry.sources)
            } else if (ts.isCallExpression(arg)) {
              const sourceFile = arg.getSourceFile();
              const position = sourceFile.getLineAndCharacterOfPosition(
                arg.getStart(),
              );
              const references = server.getReferencesAtPosition(
                sourceFile.fileName,
                arg.getStart(),
              );

              const source = {
                character: position.character,
                line: position.line,
                fileName: sourceFile.fileName,
                content: arg.getText(),
                references: [],
                modules: {},
              };
              // Immediately add the source to the entry
              // Though the source will be filled with references later
              entry.sources.push(source);

              let ref: TsReferenceEntry | undefined = references.shift();
              while (ref) {
                const currentFile = fileMap[ref.fileName];
                if (!currentFile) {
                  continue;
                }

                const refNode = findNodeAtPosition(
                  ts,
                  currentFile,
                  ref.textSpan.start,
                );
                const referencedModule = getImportTypeDefinitionFileName(
                  ts,
                  refNode,
                  program,
                );
                if (referencedModule) {
                  source.modules[referencedModule.name] = referencedModule;
                  break;
                }

                ref = references.shift();
              }
            }
          }
        }
      }

      ts.forEachChild(node, visit);
    }

    ts.forEachChild(scope, visit);
  }

  return {
    results: apps,
    errorCount,
  };
}

function createRootReferenceForArrowFunction(node: TsArrowFunction, ts: TsType): SourceReference {
  const sourceFile = node.getSourceFile();
  const position = sourceFile.getLineAndCharacterOfPosition(
    node.getStart(),
  );

  return {
    character: position.character,
    line: position.line,
    fileName: sourceFile.fileName,
    content: node.getText(),
    references: [],
    modules: {},
  }


  // return current;
}

function getImportTypeDefinitionFileName(
  ts: TsType,
  node: TsNode,
  program: TsProgram,
): ModuleReference | undefined {
  const checker = program.getTypeChecker();
  const symbol = checker.getSymbolAtLocation(node);

  const declarations = symbol?.getDeclarations();

  if (!declarations || declarations.length === 0) {
    return;
  }

  const declaration = declarations[0];
  const n = declaration.parent.parent;

  // Look for the import statement
  let nextSibling = n && getNextSibling(n);
  while (nextSibling && nextSibling.kind !== ts.SyntaxKind.StringLiteral) {
    nextSibling = getNextSibling(nextSibling);
  }

  if (nextSibling) {
    let text = nextSibling.getText();
    try {
      text = JSON.parse(text);
    } catch {
      // swallow error
    }
    const options = program.getCompilerOptions();
    const host: TsModuleResolutionHost = {
      fileExists: ts.sys.fileExists,
      readFile: ts.sys.readFile,
    };
    const result = ts.resolveModuleName(
      text,
      node.getSourceFile().fileName,
      options,
      host,
      undefined,
      undefined,
      undefined,
    );


    if (
      result.resolvedModule?.isExternalLibraryImport
    ) {
      return {
        import: node.getText(),
        importPath: text,
        name: result.resolvedModule.packageId.name,
        version: result.resolvedModule.packageId.version,
      };
    }
    if (text.startsWith("node:")) {
      return {
        import: node.getText(),
        importPath: text,
        name: text,
      }
    }
  }
}

function getNextSibling(node: TsNode): TsNode | undefined {
  const parent = node.parent;
  if (!parent) {
    // console.log('no parent');
    return undefined;
  }

  const children = parent.getChildren();
  for (let i = 0; i < children.length; i++) {
    if (children[i] === node && i + 1 < children.length) {
      return children[i + 1];
    }
  }

  return undefined;
}
function findNodeAtPosition(
  ts: TsType,
  sourceFile: TsSourceFile,
  position: number,
): TsNode | undefined {
  function find(node: TsNode): TsNode | undefined {
    if (position >= node.getStart() && position < node.getEnd()) {
      return ts.forEachChild(node, find) || node;
    }
    return undefined;
  }

  return find(sourceFile);
}

// console.log("node", node.kind, ts.SyntaxKind[node.kind], node.getText());
function inspectNode(node: TsNode, ts: TsType) {
  console.log("node", node.kind, ts.SyntaxKind[node.kind], node.getText());
  console.log("Inspecting node", {
    // isFunctionExpression: ts.isFunctionExpression(node),
    // isArrowFunction: ts.isArrowFunction(node),
    // isFunctionDeclaration: ts.isFunctionDeclaration(node),
    // isFunctionTypeNode: ts.isFunctionTypeNode(node),
    // isCallExpression: ts.isCallExpression(node),
    // isPropertyAccessExpression: ts.isPropertyAccessExpression(node),
    // isIdentifier: ts.isIdentifier(node),
    // isStringLiteral: ts.isStringLiteral(node),
    // isNumericLiteral: ts.isNumericLiteral(node),
    // isObjectLiteralExpression: ts.isObjectLiteralExpression(node),
    // isTypeLiteralNode: ts.isTypeLiteralNode(node),
    // isTypeReferenceNode: ts.isTypeReferenceNode(node),
    // isTypeQueryNode: ts.isTypeQueryNode(node),
    // isTypePredicateNode: ts.isTypePredicateNode(node),
    // // isTypeAssertion: ts.isTypeAssertion(node),
    // isAsExpression: ts.isAsExpression(node),
    // isNonNullExpression: ts.isNonNullExpression(node),
    // isNonNullChain: ts.isNonNullChain(node),
    // isParenthesizedExpression: ts.isParenthesizedExpression(node),
    // isSpreadElement: ts.isSpreadElement(node),
    // isShorthandPropertyAssignment: ts.isShorthandPropertyAssignment(node),
    // isPropertyAssignment: ts.isPropertyAssignment(node),
    // isPropertyDeclaration: ts.isPropertyDeclaration(node),
    // isPropertySignature: ts.isPropertySignature(node),
    // isAccessor: ts.isAccessor(node),
    isImportAttribute: ts.isImportAttribute(node),
    isImportAttributeName: ts.isImportAttributeName(node),
    isImportAttributes: ts.isImportAttributes(node),
    isImportClause: ts.isImportClause(node),
    isImportDeclaration: ts.isImportDeclaration(node),
    isImportEqualsDeclaration: ts.isImportEqualsDeclaration(node),
    isImportOrExportSpecifier: ts.isImportOrExportSpecifier(node),
    isImportSpecifier: ts.isImportSpecifier(node),
    isImportTypeAssertionContainer: ts.isImportTypeAssertionContainer(node),
    isImportTypeNode: ts.isImportTypeNode(node),
  });
}

// Extract references from the arrow function
function extractReferencesFromArrowFunction(
  arrowFunction: TsArrowFunction,
  program: TsProgram,
  ts: TsType,
): Array<SourceReference> {

  const rootReference = createRootReferenceForArrowFunction(arrowFunction, ts);
  const references: Array<SourceReference> = [
    rootReference,
  ];

  function visit(node: TsNode) {
    if (ts.isIdentifier(node)) {
      const dependency = getImportTypeDefinitionFileName(ts, node, program);
      if (dependency) {
        if (
          rootReference.modules[dependency.name] !== undefined
        ) {
          rootReference.modules[dependency.name].push(dependency)
        } else {
          rootReference.modules[dependency.name] = [dependency];
        }
        return;
      }
    }
    ts.forEachChild(node, visit);
  }

  visit(arrowFunction);
  return references;
}

// Function to analyze the arrow function and its references
// function analyzeSource(filePath: string, program: TsProgram, ts: TsType) {
//   const sourceFile = program.getSourceFile(filePath);

//   if (!sourceFile) {
//     throw new Error(`Cannot find source file: ${filePath}`);
//   }

//   function findArrowFunctions(node: TsNode) {
//     if (ts.isArrowFunction(node)) {
//       const references = extractReferencesFromArrowFunction(node, sourceFile, program);
//       console.log(references);
//     }
//     ts.forEachChild(node, findArrowFunctions);
//   }

//   findArrowFunctions(sourceFile);
// }
