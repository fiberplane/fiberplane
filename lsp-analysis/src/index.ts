export { Watcher } from "./watcher";
export { setupMonitoring } from "./setup";
