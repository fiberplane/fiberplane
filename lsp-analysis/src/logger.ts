import { createConsola } from "consola";

export const logger = createConsola({
  level: 4,
});
