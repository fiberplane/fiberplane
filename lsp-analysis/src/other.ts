import { Hono } from "hono";
export { Hono as App } from "hono";
// export const App = Hono;
const app = new Hono();
console.log("app", app);
// const app2 = new Hono();
// export type SuperHono = Hono & { super: boolean };
// export const App = Hono;
