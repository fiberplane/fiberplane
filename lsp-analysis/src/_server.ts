/**
 * This file contains the TypeScript language server instance and related functions.
 * It uses the `vscode-jsonrpc` library to create a connection to the language server
 * and handles the initialization, notification registration, and cleanup.
 *
 * Note that the language server is initialized in the user's project directory by spawning a child process,
 * and running `npx typescript-language-server --stdio`.
 * We interface with that process via stdin/stdout.
 */

import { type ChildProcessWithoutNullStreams, spawn } from "node:child_process";
import { resolve } from "node:path";
import {
  type MessageConnection,
  StreamMessageReader,
  StreamMessageWriter,
  createMessageConnection,
} from "vscode-jsonrpc/node.js";
import type {
  ConfigurationParams,
  DidChangeConfigurationParams,
  //   // createMessageConnection,
  InitializeParams,
} from "vscode-languageserver-protocol";
import { logger } from "./logger.js";
import { isPublishDiagnosticsParams } from "./types";
import { getFileUri } from "./utils";
// import { connect } from "node:http2";

// let tsServerInstance: {
//   connection: MessageConnection;
//   tsServer: ChildProcessWithoutNullStreams;
// } | null = null;

/**
 * Retrieves the TypeScript language server instance for a given project path.
 * If an instance already exists, it will be reused to avoid unnecessary initialization.
 *
 * @param {string} pathToProject - The path to the project directory.
 * @returns {Promise<{ connection: MessageConnection; tsServer: ChildProcessWithoutNullStreams }>} The TypeScript language server instance.
 */
export async function getTSServer(pathToProject: string) {
  const resolvedPath = resolve(pathToProject);

  // if (tsServerInstance) {
  //   // logger.debug(
  //   //   chalk.dim(
  //   //     `[debug] Reusing existing TS Server instance for project: ${resolvedPath}`,
  //   //   ),
  //   // );
  //   return tsServerInstance;
  // }

  return await initializeTSServer(resolvedPath);
  // return tsServerInstance;
}

async function initializeTSServer(pathToProject: string) {
  logger.debug(`Initializing TS Server for project: ${pathToProject}`);

  const tsServer = spawn(
    "npx",
    [
      "typescript-language-server",
      "--log-level",
      "verbose",
      // "--logVerbosity=verbose",
      "--stdio",
      // "--tsserver-log-file",
      // "out.log",
    ],
    {
      // NOTE - This will add quite a bit of startup time if the user has not yet downloaded typescript-language-server dependency before via npx...
      //        And I haven't tested what that overhead is...
      // NOTE - We *could* run the language server in the project root (via cwd switch), but I think we handle this by initializing the language server
      //        and pointing it to the project root below anyhow.
      // cwd: pathToProject,
      shell: true,
    },
  );

  // Terminate the language server when the Node.js process exits
  process.on("exit", () => {
    logger.debug(
      `[debug] Terminating TS Server instance for project: ${pathToProject}`,
    );
    tsServer.kill();
  });

  // NOTE - Uncomment to debug raw output of ts-language-server
  //
  tsServer.stdout.on("data", (data) => {
    // console.log("<TS-SERVER-DATA>");
    console.log(`--> tsServer stdout: ${data.toString()}`);
    // logger.box(`tsServer stdout: ${data.toString()}`);
    // console.log("</TS-SERVER-DATA>");
  });

  // tsServer.stderr.on("data", (data) => {
  //   logger.error(`tsserver stderr: ${data.toString()}`);
  // });

  tsServer.stdout.on("data", (data) => {
    logger.trace(`tsserver stdout: ${data.toString()}`);
  });

  const connection = createMessageConnection(
    new StreamMessageReader(tsServer.stdout),
    new StreamMessageWriter(tsServer.stdin),
  );

  connection.listen();

  tsServer.on("close", (code) => {
    logger.debug(`typescript-language-server process exited with code ${code}`);
  });

  try {
    const rootUri = getFileUri(pathToProject);
    logger.debug(
      `Initializing typescript language server with rootUri: ${rootUri}`,
    );

    // ClientCapabilities
    const params: InitializeParams = {
      processId: process.pid,
      rootUri: rootUri,
      capabilities: {
        workspace: {
          // codeLens: {
          //   // dynamicRegistration: true,
          //   refreshSupport: true,
          // },
          configuration: true,
        },
      },
      workspaceFolders: [{ uri: rootUri, name: "app" }],
      trace: "verbose",

      initializationOptions: {
        typescript: {
          referencesCodeLens: {
            enabled: true,
          },
          implementationsCodeLens: {
            enabled: true,
          },
        },

        preferences: {
          allowIncompleteCompletions: true,
          includeCompletionsForModuleExports: true,
          includeCompletionsWithInsertText: true,
        },
      },
    };

    const ir = await connection.sendRequest("initialize", params);
    console.log("ir", ir);
    await connection.sendNotification("initialized");

    registerNotificationHandlers(connection);
    await new Promise((resolve) => setTimeout(resolve, 200));
    // enable codelens
    const configParams: DidChangeConfigurationParams = {
      settings: {
        asdasdfasd: "erwe",
        // typescript: {
        referencesCodeLens: {
          enabled: true,
        },
        implementationsCodeLens: {
          enabled: true,
        },
        // },
      },
    };
    // const dcc =
    await connection.sendNotification(
      "workspace/didChangeConfiguration",
      configParams,
    );
    // console.log("configUpdate dcc ", dcc);

    // const configQueryParams: ConfigurationParams = {
    //   items: [
    //     {
    //       section: "typescript.referencesCodeLens",
    //     },
    //     {
    //       section: "typescript.implementationsCodeLens",
    //     },
    //   ],
    // };
    // const r = await connection.sendRequest(
    //   "workspace/configuration",
    //   configQueryParams,
    // );
    // console.log("configUpdate", r);

    return { connection, tsServer };
  } catch (error) {
    logger.error("Error initializing TS Server:", error);
    throw error;
  }
}

/**
 * Registers handlers for various notifications from the TypeScript language server.
 *
 * @param {MessageConnection} connection - The connection to the TypeScript language server.
 */
function registerNotificationHandlers(connection: MessageConnection) {
  const logger = console;
  connection.onNotification("textDocument/didOpen", (params) => {
    logger.debug(
      `[textDocument/didOpen] Received definition request: ${JSON.stringify(
        params,
        null,
        2,
      )}`,
    );
  });
  // Listen for diagnostics
  connection.onNotification(
    "textDocument/publishDiagnostics",
    (params: unknown) => {
      if (!isPublishDiagnosticsParams(params)) {
        logger.debug(
          "[debug] Received unexpected params for `textDocument/publishDiagnostics`",
        );
        return;
      }
      const { uri, diagnostics } = params;
      if (diagnostics.length > 0) {
        logger.info(
          `[textDocument/publishDiagnostics] Diagnostics for ${uri}:`,
        );
        for (const diag of diagnostics) {
          logger.info(
            `- [${diag.severity}] ${diag.message} at ${diag.range.start.line}:${diag.range.start.character}`,
          );
        }
      } else {
        logger.debug(
          `[textDocument/publishDiagnostics] No diagnostics for ${uri}.`,
        );
      }
    },
  );

  // Listen for log messages
  connection.onNotification("window/logMessage", (params) => {
    const { type, message } = params;
    logger.info(`[window/logMessage] Log Message [${type}]: ${message}`);
    logger.debug(params);
  });

  // Listen for show messages
  connection.onNotification("window/showMessage", (params) => {
    const { type, message } = params;
    logger.debug(`[window/showMessage] Show Message [${type}]: ${message}`);
  });
}

// type
