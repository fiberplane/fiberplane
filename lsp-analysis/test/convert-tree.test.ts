import path from "node:path";
import { assert, expect, test } from "vitest";
import { AppFactory, analyze, setupMonitoring } from "../src";
// import type { RouteTree } from "../src/types";

// test.skip("simple tree", async () => {
//   const trees: Array<RouteTree> = [
//     {
//       type: "ROUTE_TREE",
//       baseUrl: "",
//       id: "root",
//       entries: [
//         {
//           type: "ROUTE_ENTRY",
//           id: "home",
//           method: "get",
//           path: "/",
//           sources: [],
//         },
//       ],
//       fileName: "index.ts",
//       name: "app",
//     },
//   ];

//   const factory = new AppFactory(trees);
//   const app = factory.getApp("root");
//   expect(factory.getHistoryLength()).toBe(0);
//   const request = new Request("http://localhost/", { method: "GET" });
//   const response = await app.fetch(request);
//   expect(await response.text()).toEqual("Ok");
//   expect(factory.getHistoryLength()).toBe(2);
//   expect(factory.hasVisited("root")).toBe(true);
//   expect(factory.hasVisited("home")).toBe(true);
// });

// test.skip("nested tree", async () => {
//   const trees: Array<RouteTree> = [
//     {
//       type: "ROUTE_TREE",
//       baseUrl: "",
//       id: "root",
//       entries: [
//         {
//           type: "ROUTE_TREE_REFERENCE",
//           targetId: "nested",
//           fileName: "nested.ts",
//           name: "nested",
//           path: "/",
//         },
//       ],
//       fileName: "index.ts",
//       name: "app",
//     },
//     {
//       type: "ROUTE_TREE",
//       baseUrl: "",
//       id: "nested",
//       entries: [
//         {
//           type: "ROUTE_ENTRY",
//           id: "nested-home",
//           method: "get",
//           path: "/nested",
//           sources: [],
//         },
//       ],
//       fileName: "nested.ts",
//       name: "nested",
//     },
//   ];

//   const factory = new AppFactory(trees);
//   const app = factory.getApp("root");
//   expect(factory.getHistoryLength()).toBe(0);

//   // Make request
//   const request = new Request("http://localhost/nested", { method: "GET" });
//   const response = await app.fetch(request);

//   expect(await response.text()).toEqual("Ok");
//   expect(factory.getHistoryLength()).toBe(3);
//   expect(factory.hasVisited("root")).toBe(true);
//   expect(factory.hasVisited("nested")).toBe(true);
//   expect(factory.hasVisited("nested-home")).toBe(true);
// });

// test.skip("nested tree", async () => {
//   const trees: Array<RouteTree> = [
//     {
//       type: "ROUTE_TREE",
//       baseUrl: "",
//       id: "root",
//       entries: [
//         {
//           type: "MIDDLEWARE_ENTRY",
//           id: "all-middleware",
//           path: "",
//           sources: [],
//         },
//         {
//           type: "ROUTE_ENTRY",
//           id: "subpath",
//           method: "get",
//           path: "/subpath",
//           sources: [],
//         },
//         {
//           type: "MIDDLEWARE_ENTRY",
//           id: "subpath-id-middleware",
//           path: "/subpath/:id",
//           sources: [],
//         },
//         {
//           type: "ROUTE_ENTRY",
//           id: "subpath-id",
//           method: "get",
//           path: "/subpath/:id",
//           sources: [],
//         },
//       ],
//       fileName: "index.ts",
//       name: "app",
//     },
//   ];

//   const factory = new AppFactory(trees);
//   const app = factory.getApp("root");
//   expect(factory.getHistoryLength()).toBe(0);

//   // Make request
//   let request = new Request("http://localhost/subpath", { method: "GET" });
//   let response = await app.fetch(request);

//   expect(await response.text()).toEqual("Ok");
//   // The middleware should be called for all requests
//   expect(factory.hasVisited("root")).toBe(true);
//   expect(factory.hasVisited("subpath")).toBe(true);
//   expect(factory.hasVisited("all-middleware")).toBe(true);
//   // That's 3 in total
//   expect(factory.getHistoryLength()).toBe(3);

//   // Reset the history
//   factory.resetHistory();

//   // Make another request
//   request = new Request("http://localhost/subpath/:id", { method: "GET" });
//   response = await app.fetch(request);

//   expect(await response.text()).toEqual("Ok");

//   expect(factory.hasVisited("root")).toBe(true);
//   expect(factory.hasVisited("subpath-id")).toBe(true);
//   expect(factory.hasVisited("all-middleware")).toBe(true);
//   expect(factory.hasVisited("subpath-id-middleware")).toBe(true);
//   expect(factory.getHistoryLength()).toBe(4);
// });
test("hono-factory", async () => {
  const absolutePath = path.join(__dirname, "./test-case/hono-factory");
  const { watcher, findHonoRoutes, teardown } = setupMonitoring(absolutePath);
  try {
    watcher.start();
    const start = performance.now();
    const result = findHonoRoutes();
    console.log(
      `Duration for hono-factory: ${(performance.now() - start).toFixed(4)}`,
    );
    const { resourceManager } = result;
    const root = analyze(resourceManager.getResources());
    expect(root).not.toBeNull();
    assert(root !== null);
    const factory = new AppFactory(resourceManager);
    if (!root) {
      throw new Error("Root is null");
    }

    const app = factory.getApp(root.id);
    const request = new Request("http://localhost/", { method: "GET" });
    const response = await app.fetch(request);
    expect(await response.text()).toEqual("Ok");
    expect(factory.getHistoryLength()).toBe(3);

    const history = factory.getHistory();
    expect(factory.hasVisited(root.id)).toBeTruthy();
    // Expect the app from the factory file to be visited
    expect(
      history.find((item) => item.startsWith("ROUTE_TREE:factory.ts@")),
    ).toBeTruthy();
    expect(
      history.find((item) => item.startsWith("ROUTE_ENTRY:factory.ts@")),
    ).toBeTruthy();
    expect(factory.getFilesForHistory()).toMatchSnapshot();
  } finally {
    teardown();
  }
});

test("barrel-files", async () => {
  const absolutePath = path.join(__dirname, "./test-case/barrel-files");
  const { watcher, findHonoRoutes, teardown } = setupMonitoring(absolutePath);
  try {
    watcher.start();
    // findHonoRoutes();
    const start = performance.now();
    const result = findHonoRoutes();
    console.log(
      `Duration for barrel files: ${(performance.now() - start).toFixed(4)}`,
    );
    const root = analyze(result.resourceManager.getResources());
    expect(root).not.toBeNull();
    assert(root !== null);
    const factory = new AppFactory(result.resourceManager);
    if (!root) {
      throw new Error("Root is null");
    }

    const app = factory.getApp(root.id);
    const request = new Request("http://localhost/", { method: "GET" });
    const response = await app.fetch(request);
    expect(await response.text()).toEqual("Ok");
    expect(factory.getHistoryLength()).toBe(2);

    expect(factory.hasVisited(root.id)).toBeTruthy();
    // Expect the app from the factory file to be visited
    expect(factory.getFilesForHistory()).toMatchSnapshot();
  } finally {
    teardown();
  }
});
