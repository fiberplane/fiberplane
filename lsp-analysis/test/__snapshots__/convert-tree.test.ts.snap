// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`barrel-files 1`] = `
"//@ Start file: index.ts
import { cors } from "hono/cors";

const app = new Hono();
app.baseUrl = "";
          
// Route entry:ROUTE_ENTRY:index.ts@407
app.get("/user/1", cors()async (c) => {
  // await getUser();
  const user = await getUser();
  return c.json(user);
}
// Route entry:ROUTE_ENTRY:index.ts@531
app.get("/", (c) => c.text("Hello, Hono!")
//@ EOF file: index.ts


//@ Start file: db/index.ts


// Source reference:SOURCE_REFERENCE:db%2Findex.ts@9
export { getUser } from "./getUser"
//@ EOF file: db/index.ts


//@ Start file: db/getUser.ts
import { measure } from "@fiberplane/hono-otel";

// Source reference:SOURCE_REFERENCE:db%2FgetUser.ts@128
const DEFAULT_USER_NAME = "John Doe"
// Source reference:SOURCE_REFERENCE:db%2FgetUser.ts@166
const DEFAULT_EMAIL = "john@doe.com"
// Source reference:SOURCE_REFERENCE:db%2FgetUser.ts@222
const getUser = measure("getUser", async () => {
  await sleep(100);
  const value: User = {
    name: DEFAULT_USER_NAME,
    email: DEFAULT_EMAIL,
  };
  return value;
})
//@ EOF file: db/getUser.ts


//@ Start file: utils.ts


// Source reference:SOURCE_REFERENCE:utils.ts@0
export async function sleep(duration: number) {
  return new Promise((resolve) => setTimeout(resolve, duration));
}
//@ EOF file: utils.ts


//@ Start file: types.ts


// Source reference:SOURCE_REFERENCE:types.ts@0
export type Email = string;
// Source reference:SOURCE_REFERENCE:types.ts@29
export type User = {
  name: string;
  email: Email;
};
//@ EOF file: types.ts
"
`;

exports[`hono-factory 1`] = `
"//@ Start file: index.ts


const subHello = new Hono();
subHello.baseUrl = "";
          
// Route entry:ROUTE_ENTRY:index.ts@124
subHello.get("/", (c) => c.text("Hello, sub!")
// Route entry:ROUTE_ENTRY:index.ts@173
subHello.get("/bye", (c) => c.text("Bye, sub!")
const app = new Hono();
app.baseUrl = "";
          
//@ EOF file: index.ts


//@ Start file: factory.ts


const app = new Hono();
app.baseUrl = "";
          
// Route entry:ROUTE_ENTRY:factory.ts@281
app.get("/", (c) => c.text("Hello, Hono!")
//@ EOF file: factory.ts


//@ Start file: bye.ts


const bye = new Hono();
bye.baseUrl = "";
          
// Route entry:ROUTE_ENTRY:bye.ts@54
bye.get("/bye", (c) => c.text("Bye, sub!")
//@ EOF file: bye.ts


//@ Start file: silence.ts


const silence = new Hono();
silence.baseUrl = "";
          
// Route entry:ROUTE_ENTRY:silence.ts@58
silence.get("/", (c) => c.text("...")
//@ EOF file: silence.ts


//@ Start file: panic.ts


const app = new Hono();
app.baseUrl = "";
          
// Route entry:ROUTE_ENTRY:panic.ts@54
app.get("/", (c) => {
  c.status(500);
  return c.text("Panic!");
}
//@ EOF file: panic.ts
"
`;
